File: Python/LinkedList/DoublyLinkedList.py
#Python Implementation Doubly Linked Lists

class Node:

    def __init__(self, data):
        self.dataval = data
        self.nextval = None
        self.preval = None

class DLinkedList:

    def __init__(self):
        self.head = None

    def append(self, data):
        if self.head is None:
            self.head = Node(data)
            self.lastnode = self.head
        else:
            temp = self.lastnode
            self.lastnode.nextval = Node(data)
            self.lastnode = self.lastnode.nextval
            self.lastnode.preval = temp

    def prepend(self,data):
        if self.head is None:
            self.head = Node(data)
        else:
            new_node = self.head
            new_node.preval = Node(data)
            self.head = new_node.preval
            self.head.nextval = new_node

    def removeAtStart(self):
        if self.head is None:
            return ('The Linked List is empty')
        
        temp = self.head
        self.head = self.head.nextval
        self.head.preval = None
        del temp

    def insert(self,data,index):
        curr_node = self.head
        i = 1

        while i < index  and curr_node != None:
            curr_node = curr_node.nextval
            i+=1

        if index == 0:
            self.prepend(data)
        else:
            new_node = Node(data) 
            new_node.nextval = curr_node.nextval
            temp = curr_node.nextval
            temp.preval = new_node
            curr_node.nextval = new_node
            new_node.preval = curr_node

    def delete(self, index):
        curr_node = self.head
        i = 0

        while i < index and curr_node != None:
            curr_node = curr_node.nextval
            i+=1
        
        if(index == 0):
            self.removeAtStart()
        else:
            temp = curr_node.preval
            temp.nextval = curr_node.nextval
            temp = curr_node.nextval
            temp.preval = curr_node.preval
            del curr_node

    def reverse(self):
        next_node = self.head
        temp = None

        if self.head.nextval is None:
            return None

        else:
            while next_node is not None:
                temp = next_node.preval
                next_node.preval = next_node.nextval
                next_node.nextval = temp
                next_node = next_node.preval


            self.head.nextval = None
            self.head = temp.preval
            

    def displayAll(self):
        curr_node = self.head
        if self.head is None:
            print("Empty Linked List!")
        else:
            while curr_node is not None:
                print(curr_node.dataval, end=" ")
                curr_node = curr_node.nextval

myLinkedList = DLinkedList()

myLinkedList.append(5)

myLinkedList.append(10)

myLinkedList.append(22)

myLinkedList.append(7)

myLinkedList.prepend(26)

myLinkedList.prepend(13)

myLinkedList.insert(17, 2)

myLinkedList.delete(2)

myLinkedList.insert(33, 4)

print('Linked List:', end=" ")

myLinkedList.displayAll()

print('')

print('Reversed Linked List:', end=" ")

myLinkedList.reverse()

myLinkedList.displayAll()

File: C++/LinkedList/DoublyLinkedList.cpp
#include <iostream>
#include <cstdlib>
using namespace std;

struct Node {
    int value;
    struct Node * next;
    struct Node * prev;
};

class LinkedList
{
    public:

        LinkedList()
        {
            head = nullptr;
            tail = nullptr;
            length = 1;
        }

        // Add new value into the linked list
        void append(int value)
        {
            Node *new_node = new Node;
            new_node->value = value;
            new_node->next = nullptr;
            new_node->prev = nullptr;

            if (head == nullptr)
            {
                head = new_node;
                tail = new_node;
            }
            else
            {	new_node->prev = tail;
                tail->next = new_node;
                tail = new_node;
            }
            length++;
        }

        // Add new value to the first of linked list
        void prepend(int value) {
            Node * new_node;
            new_node = new Node;

            new_node->value = value;
            new_node->next = head;
            head->prev = new_node;

            head = new_node;
            length++;
        }

        // Print all the linked list value
        void printList() {
            Node * current_node = head;

            while (current_node != nullptr) {
                cout << "\r" <<current_node->value << ",";
                current_node = current_node->next;
            }
        }

        void insert(int index, int value){
            // Check for proper parameters
            if(index >= length){
                cout << "yes" << endl;
                return append(value);
            }

            Node * new_node = new Node;
            new_node->value = value;
            new_node->next = nullptr;
            Node * leader = traverseToIndex(index-1);
            Node * follower = leader->next;
            leader->next = new_node;
            new_node->prev = leader;
            new_node->next = follower;
            follower->prev = new_node;
            length++;

        }

        Node * traverseToIndex(int index){
            // Check parameters
            int counter = 0;
            Node * current_node = head;
            while(counter != index){
                current_node = current_node->next;
                counter++;
            }

            return current_node;
        }

        // Remote first value
        int pop() {
            int retval = 0;
            Node * next_node = nullptr;

            next_node = head->next;
            retval = head->value;            
            delete head;
            head = next_node;
            head->prev = nullptr;
            length--;

            return retval;
        }

        void remove(int index){
            // Check parameters
            Node * leader = traverseToIndex(index-1);
            Node * unwantedNode = leader->next;
            Node * follower = unwantedNode->next;
            leader->next = follower;
            follower->prev = leader;
            delete unwantedNode;
            length--;
        }

        void reverse(){
            if(head->next == nullptr){
                return;
            }
            Node * first = head;
            tail = head;
            Node * second = first->next;

            while(second != nullptr){
                Node * temp = second->next;
                second->next = first;
                first->prev = second;
                first = second;
                second = temp;
            }

            head->next = nullptr;
            head = first;

            return printList();
        }

    private:
        Node *head;
        Node *tail;
        int length;
};

int main() {

    LinkedList linkedlist;

    linkedlist.append(5);
    linkedlist.append(16);
    linkedlist.prepend(1);
    linkedlist.printList();
    cout << "\n--------------" << endl;
    linkedlist.insert(2, 99);
    linkedlist.printList();
    cout << "\n--------------" << endl;
    linkedlist.remove(2);
    linkedlist.printList();
    cout << "\n--------------" << endl;
    linkedlist.pop();
    linkedlist.printList();
    cout << "\n--------------" << endl;
    linkedlist.reverse();

    return 0;
}